sdks/python/apache_beam/runners/interactive/pipeline_fragment_test.py

```diff
@@ -129,6 +129,41 @@ def test_fragment_does_not_prune_teststream(self):
     # resulting graph is invalid and the following call will raise an exception.
     fragment.to_runner_api()
 
+  @patch('IPython.get_ipython', new_callable=mock_get_ipython)
+  def test_pipeline_composites(self, cell):
+    """Tests that composites are supported.
+    """
+    with cell:  # Cell 1
+      p = beam.Pipeline(ir.InteractiveRunner())
+      ib.watch({'p': p})
+
+    with cell:  # Cell 2
+      # pylint: disable=range-builtin-not-iterating
+      init = p | 'Init' >> beam.Create(range(5))
+
+    with cell:  # Cell 3
+      # Have a composite within a composite to test that all transforms under a
+      # composite are added.
+
+      @beam.ptransform_fn
+      def Bar(pcoll):
+        return pcoll | beam.Map(lambda n: n)
+
+      @beam.ptransform_fn
+      def Foo(pcoll):
+        p1 = pcoll | beam.Map(lambda n: n)
+        p2 = pcoll | beam.Map(str)
+        bar = pcoll | Bar()
+        return {'pc1': p1, 'pc2': p2, 'bar': bar}
+
+      res = init | Foo()
+
+    ib.watch(locals())
+    pc = res['pc1']
+
+    result = pf.PipelineFragment([pc]).run()
+    self.assertEqual([0, 1, 2, 3, 4], list(result.get(pc)))
```

Comment:
I'm not sure this is working as intended. I tried changing this test like so:

```py
      @beam.ptransform_fn
      def Bar(pcoll):
        return pcoll | beam.Map(lambda n: 2*n)

      @beam.ptransform_fn
      def Foo(pcoll):
        p1 = pcoll | beam.Map(lambda n: 3*n)
        p2 = pcoll | beam.Map(str)
        bar = p1 | Bar()
        return {'pc1': p1, 'pc2': p2, 'bar': bar}

      res = init | Foo()
      #pc = init | beam.Map(lambda n: 6*n)

    ib.watch(locals())
    pc = res['bar']

    result = pf.PipelineFragment([pc]).run()
    self.assertEqual([0, 6, 12, 18, 24], list(result.get(pc)))
```

(note the 2x, 3x, to make sure pcollection values are different)

and the test fails the assertion

Reply:
I tried it and that's true. Investigating if it's an issue with the test setup or a bug...

Reply:
@TheNeuralBit This should be a bug in the test itself. `ib.watch()` doesn't resolve nested dictionaries. So `ib.watch(locals())` is a NOOP here. It should be `ib.watch(res)` directly.

There is another bug that when a cache key is not found for a PCollection, an empty str '' is returned for cache manager to match cache files, where '' is matched to anything. This is a severe bug but only occurs when using the PipelineResult's get API (not encouraged) and PCollections are not correctly watched. I'll send a fix for it.

Reply:
Sent https://github.com/apache/beam/pull/16440, PTAL, thx!

Reply:
Awesome, thanks for the quick turnaround!